### PR TITLE
[#11450] 'Session opening soon' alerts to instructors: Also give the join link if applicable

### LINKED ISSUE #11450: 'Session opening soon' alerts to instructors: Also give the join link if applicable
Current: When a session is due to open soon, TEAMMATES sends an alert to instructors of that course 24 hours prior to the opening time. It contains a link to edit the session.

Example:
image

Problem: The given link will not work if the instructor has not joined the course yet.

Suggestion: If the instructor hasn't joined the course yet, give the join link as well, with an appropriate explanation.

### PR DESCRIPTION
**Pull request by josiahkhoo**
Fixes #11450

**Outline of Solution**

Added more instructions 'Session opening soon' alerts if the instructor has not joined the course yet (created an account). 

This is the original email (instructors who joined the course will still receive this):

image

This is the updated email for instructors who have yet to join the course:

image

### CODE REVIEW

**Comment by josiahkhoo** [comment]
@damithc adopted your suggestion and updated the copy for the email to this:

image

**Comment by damithc** [comment]
@josiahkhoo there are two options:
1. Use the same email template for both emails, but vary parts based on join status. Pros: less duplication. 
2. Use two separate email templates . Pros: can optimize the email for each case

Which one are you using? Which one _should_ we use (given the pros/cons)?

**Comment by damithc** [comment]
> I created two separate fragments for the bottom section based on join status. Feel that the first option is more suitable since less duplication makes this more maintainable. At the same time, it still allows the copy to be optimized to each case (except for the top section).

@josiahkhoo Got it. I'm fine with the current version of the email.

File: src/main/resources/ownerEmailFragment-editDetails.html [hunk 1]
```diff
@@ -0,0 +1,10 @@
+<p>
+<ul>
+    <li>
+        <strong>To edit details of this session,</strong> please visit <a href="${sessionEditUrl}">${sessionEditUrl}</a>
+    </li>
+    <li>
+        No other action is needed if you deem the session details to be correct.
+    </li>
```
Comment (wkurniawan07) [inline src/main/resources/ownerEmailFragment-editDetails.html:8]: Use 2-space indentation for HTML files

File: src/main/resources/ownerEmailFragment-joinCourseBeforeEditDetails.html [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    <li>
+        If prompted to log in when joining the course, use your Google account to log in.
+        If you do not have a Google account, please create one from the <a href="https://accounts.google.com/NewAccount">Google Accounts page</a>.
+    </li>
+    <li>
+        The above link to join the course is unique to you. Please do not share it with others.
+    </li>
+    <li>
+        No other action is needed if you deem the session details to be correct.
+    </li>
```
Comment (wkurniawan07) [inline src/main/resources/ownerEmailFragment-joinCourseBeforeEditDetails.html:17]: Similarly here

File: src/test/java/teammates/logic/api/EmailGeneratorTest.java [hunk 3]
```diff
@@ -217,14 +217,23 @@ public void testGenerateFeedbackSessionEmails() throws Exception {
...
 
-        assertTrue(coOwner1.hasCoownerPrivileges());
+        assertTrue(coOwnerNotJoined.hasCoownerPrivileges());
 
-        verifyEmailReceivedCorrectly(emails, coOwner1.getEmail(), subject,
-                "/sessionOpeningSoonEmailForCoOwner.html");
+        verifyEmailReceivedCorrectly(emails, coOwnerNotJoined.getEmail(), subject,
+                "/sessionOpeningSoonEmailForCoOwnerNotJoined.html");
+
+        // this instructor email has been given co-owner privileges in the test file but has not joined
```
Comment (wkurniawan07) [inline src/test/java/teammates/logic/api/EmailGeneratorTest.java:229]: This instructor has joined, though?

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]
Implementation looks ok; simple, clean way to support the feature.

File: src/test/java/teammates/logic/api/EmailGeneratorTest.java (hunk 3 above)
Comment (josiahkhoo) [reply src/test/java/teammates/logic/api/EmailGeneratorTest.java]: My bad, meant for this to say the instructor has joined. Made the modifications.

File: src/main/resources/ownerEmailFragment-joinCourseBeforeEditDetails.html (hunk 2 above)
Comment (josiahkhoo) [reply src/main/resources/ownerEmailFragment-joinCourseBeforeEditDetails.html]: Updated

File: src/main/resources/ownerEmailFragment-editDetails.html (hunk 1 above)
Comment (josiahkhoo) [reply src/main/resources/ownerEmailFragment-editDetails.html]: Updated

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]

File: src/main/java/teammates/logic/api/EmailGenerator.java [hunk 4]
```diff
@@ -878,12 +904,17 @@ private String fillUpStudentRejoinAfterGoogleIdResetFragment(StudentAttributes s
                 "${supportEmail}", Config.SUPPORT_EMAIL);
     }
 
-    private String fillUpInstructorJoinFragment(InstructorAttributes instructor) {
-        String joinUrl = Config.getFrontEndAppUrl(instructor.getRegistrationUrl()).toAbsoluteString();
+    private String getInstructorCourseJoinUrl(InstructorAttributes instructor) {
+        return Config.getFrontEndAppUrl(Const.WebPageURIs.JOIN_PAGE)
+                .withRegistrationKey(instructor.getKey())
+                .withEntityType(Const.EntityType.INSTRUCTOR)
+                .toAbsoluteString();
```
[the remark this answers is not in the document]
Comment (josiahkhoo) [reply src/main/java/teammates/logic/api/EmailGenerator.java]: Updated to use `Config.getFrontEndAppUrl(instructor.getRegistrationUrl()).toAbsoluteString()` instead

**Review by wkurniawan07 — APPROVED** [review]
LGTM

**Review by madanalogy — APPROVED** [review]

**Pull request merged**